File: backend/internal/database/models/opportunities.go
package models

import "time"

type Opportunity struct {
	// Unique opportunity identifier
	ID uint `json:"id" gorm:"primaryKey;autoIncrement" example:"1"`
	// Opportunity title
	Title string `json:"title" gorm:"type:varchar(255);not null" example:"AI Grant"`
	// Opportunity type
	Type string `json:"type" gorm:"type:varchar(100);not null;index:idx_opportunities_type" example:"grant"`
	// Issuing organization
	Organism string `json:"organism" gorm:"type:varchar(255);not null" example:"EU"`
	// Short description
	Description *string `json:"description,omitempty" gorm:"type:text" example:"Funding program for AI"`
	// Eligibility criteria
	Criteria *string `json:"criteria,omitempty" gorm:"type:text" example:"Student or Startup < 3 years"`
	// External reference URL
	ExternalLink *string `json:"external_link,omitempty" gorm:"type:varchar(500)" format:"uri" example:"https://example.com/opportunity"`
	// Deadline (UTC)
	Deadline *time.Time `json:"deadline,omitempty" gorm:"index:idx_opportunities_deadline" format:"date-time"`
	// Creation timestamp (UTC)
	CreatedAt time.Time `json:"created_at" format:"date-time"`
	// Update timestamp (UTC)
	UpdatedAt time.Time `json:"updated_at" format:"date-time"`
}

File: web/src/lib/api/contracts/opportunities.ts
export type OpportunityDTO = {
  id: number;
  title: string;
  type: string;
  organism: string;
  description?: string | null;
  criteria?: string | null;
  external_link?: string | null;
  deadline?: string | null;
  created_at: string;
  updated_at: string;
};

export type Opportunity = {
  id: number;
  title: string;
  type: string;
  organism: string;
  description?: string | null;
  criteria?: string | null;
  externalLink?: string | null;
  deadline?: string | null;
  createdAt: Date;
  updatedAt: Date;
};

export const mapOpportunity = (dto: OpportunityDTO): Opportunity => ({
  id: dto.id,
  title: dto.title,
  type: dto.type,
  organism: dto.organism,
  description: dto.description,
  criteria: dto.criteria,
  externalLink: dto.external_link,
  deadline: dto.deadline,
  createdAt: new Date(dto.created_at),
  updatedAt: new Date(dto.updated_at),
});
